### PR TITLE
Added possibility to remove nodes per labels

### PR DESCRIPTION
**Pull request by wadael**
apoc.remove.nodes.withlabel("label1","label2",....) will remove the nodes that one at least one of those labels.

Under the hood, it does N times
CALL apoc.periodic.commit('MATCH (n:`label`) DETACH DELETE n;',null)

### CODE REVIEW

**Comment by wadael** [comment]
Relates to https://github.com/neo4j-contrib/neo4j-apoc-procedures/issues/729

**Comment by wadael** [comment]
Works on my machine :)
Errors are in BoltTest, not impacted by my PR

**Comment by jexp** [comment]
Can you change it to withLabels
and the usage of periodic commit is wrong you have to return a count of deleted entities and use a WITH {limit} parameter within the query
`MATCH (n:'+label+') WITH n LIMIT {limit} DETACH DELETE n RETURN count(*)`

**Comment by wadael** [comment]
I was surprised that using WITH n LIMIT {limit}, not forgetting to put limit with a value in the passed params did not work. Query executed. Nodes were not deleted. I'll try to add a testcase later.

**Comment by wadael** [comment]
Other tests failing although not impacted by this PR ... grrr

File: src/main/java/apoc/nodes/Nodes.java [hunk 1]
```diff
@@ -1,14 +1,9 @@
 package apoc.nodes;
 
-import apoc.result.LongResult;
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: please don't change the imports that makes merges more complicated

File: src/main/java/apoc/nodes/Nodes.java [hunk 2]
```diff
@@ -303,6 +299,39 @@ public boolean isDense(@Name("node") Node node) {
         }
     }
 
+    @Procedure(value = "apoc.remove.nodes.withlabels", mode = Mode.WRITE)
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: don't use remove as namespace, before nodes, but instead just call this `apoc.nodes.remove`

File: src/main/java/apoc/nodes/Nodes.java [hunk 3]
```diff
@@ -303,6 +299,39 @@ public boolean isDense(@Name("node") Node node) {
...
+    @Description("apoc.remove.nodes.withlabels(['Person','Movie'],-1) | Will remove all the nodes having one of the given labels, via DETACH DELETE. Use a negative limit for no limit in deletion. Same limit used for each label.")
+    public Stream<MapResult> removeNodesWithLabels(@Name("labels") List<String> labels, @Name("limit") Long limit) {
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: Limit should be a `long`primitive and have a defaultValue entry of "-1"

File: src/main/java/apoc/nodes/Nodes.java [hunk 4]
```diff
@@ -303,6 +299,39 @@ public boolean isDense(@Name("node") Node node) {
...
+        StringBuilder query = new StringBuilder();
+
+        Map<String,Object> results = new HashMap<String, Object>();
+        HashMap<String, Object> parameters = new HashMap<>();
+
+        for( String label: labels) {
+            if(label != null && !label.trim().equals("")) {
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: !label.trim("").isEmpty()

File: src/main/java/apoc/nodes/Nodes.java [hunk 5]
```diff
@@ -303,6 +299,39 @@ public boolean isDense(@Name("node") Node node) {
...
+                query.setLength(0);
+                query.append("MATCH (n:");
+                query.append(label);
+                query.append(")");
+                if (limit > 0) {
+                    query.append(" WITH n LIMIT ");
+                    query.append(limit);
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: limit should be a parameter

File: src/main/java/apoc/nodes/Nodes.java [hunk 6]
```diff
@@ -303,6 +299,39 @@ public boolean isDense(@Name("node") Node node) {
...
+                }
+                query.append(" DETACH DELETE n RETURN");
+                query.append(" \"");
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: you don't need to return things, you can just use it from the query result statistics.

File: src/main/java/apoc/nodes/Nodes.java [hunk 7]
```diff
@@ -303,6 +299,39 @@ public boolean isDense(@Name("node") Node node) {
...
+        HashMap<String, Object> parameters = new HashMap<>();
+
+        for( String label: labels) {
+            if(label != null && !label.trim().equals("")) {
+                query.setLength(0);
+                query.append("MATCH (n:");
+                query.append(label);
+                query.append(")");
+                if (limit > 0) {
+                    query.append(" WITH n LIMIT ");
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: I think stringbuilder is overkill here, just use string.format
and make sure to use backticks arround the label

File: src/test/java/apoc/nodes/NodesTest.java [hunk 8]
```diff
@@ -175,4 +174,74 @@ public void testDegreeDirectionOnly() {
...
+                    r -> {
+                    });
+        } catch (Exception e) {
+            caught = true;
+        }
+        Assert.assertTrue(caught);
+    }
+
+    @Test
+    public void testRemoveNodesWithLabels_OkParamsWithLimit() {
```
Comment (jexp) [inline src/test/java/apoc/nodes/NodesTest.java]: also test label with spaces

File: src/test/java/apoc/nodes/NodesTest.java [hunk 9]
```diff
@@ -175,4 +174,74 @@ public void testDegreeDirectionOnly() {
...
+                params,
+                r -> {
+                    Map m = r.next();
+                    Map map = (Map) m.get("value");
+                    assertNotNull(map);
+                    assertEquals(3,map.size());
+                });
+    }
+
+    @Test
```
Comment (jexp) [inline src/test/java/apoc/nodes/NodesTest.java]: just add an expected exception to `@Test`

File: src/test/java/apoc/nodes/NodesTest.java [hunk 10]
```diff
@@ -175,4 +174,74 @@ public void testDegreeDirectionOnly() {
...
+    public void testRemoveNodesWithLabels_OkParamsWithLimit() {
+        String[] labels = {"aa", "bb", "cc"};
+        Map params = new HashMap();
+        params.put("labels", labels);
+        params.put("limit", 50);
+        TestUtil.testResult(db,
+                "CALL apoc.remove.nodes.withlabels",
+                params,
+                r -> {
+                    Map m = r.next();
```
Comment (jexp) [inline src/test/java/apoc/nodes/NodesTest.java]: use generics

File: src/main/java/apoc/nodes/Nodes.java (hunk 1 above)
Comment (jexp) [reply src/main/java/apoc/nodes/Nodes.java]: ping

**Comment by jexp** [comment]
@wadael do you want to continue on this one?

File: src/main/java/apoc/nodes/Nodes.java (hunk 6 above)
Comment (wadael) [reply src/main/java/apoc/nodes/Nodes.java]: As this is thought as a command to help with messed "load csv"s , I'd rather return the count of deleted nodes per label.

**Comment by wadael** [comment]
Hi Michael,
The conflict is the new code here.
Took time to push that Im so lame at git, had to reclone

**Comment by jexp** [comment]
There is a new conflict. Sorry. If you get it in soon today it will be part of the relase..

**Comment by wadael** [comment]
idk, my tests are after the pre existing ones. Why is there a conflict ...

File: src/main/java/apoc/nodes/Nodes.java [hunk 11]
```diff
@@ -303,6 +304,26 @@ public boolean isDense(@Name("node") Node node) {
...
+    public Stream<MapResult> removewithlabels(@Name("labels") List<String> labels, @Name(value = "limit",defaultValue="100")  Long limit) {
+        if (limit <= 0)
+            throw new RuntimeException("limit must be a positive integer");
+
+        Map<String,Object> results = new HashMap<String, Object>();
+        HashMap<String, Object> parameters = new HashMap<>();
+
+        for( String label: labels) {
+            if(label.trim().length()>0) {
+                String query = String.format("call apoc.periodic.iterate( 'MATCH (n:`%s`) RETURN n',  'DETACH DELETE n', {batchsize:%d,iterateList:true, parallel:true}  )", label, limit);
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: parallel true doesn't work here
best to prefix the statement with `cypher runtime=slotted MATCH (...)`
quote the labels

File: src/main/java/apoc/nodes/Nodes.java [hunk 12]
```diff
@@ -1,14 +1,12 @@
 package apoc.nodes;
 
+import apoc.*;
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: don't modify the imports

File: src/main/java/apoc/nodes/Nodes.java [hunk 13]
```diff
@@ -303,6 +304,26 @@ public boolean isDense(@Name("node") Node node) {
         }
     }
 
+    @Procedure(mode = Mode.WRITE)
+    @Description("apoc.nodes.removewithlabels(['Person','Movie'],1000) | Will remove all the nodes having one of the given labels, via DETACH DELETE using apoc.periodic.iterate. Same limit used for each label. Default is 100. ")
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: please change the name to removeByLabel

File: src/main/java/apoc/nodes/Nodes.java [hunk 14]
```diff
@@ -303,6 +304,26 @@ public boolean isDense(@Name("node") Node node) {
...
+    public Stream<MapResult> removewithlabels(@Name("labels") List<String> labels, @Name(value = "limit",defaultValue="100")  Long limit) {
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: make limit a `long` no boxing

File: src/main/java/apoc/nodes/Nodes.java [hunk 15]
```diff
@@ -303,6 +304,26 @@ public boolean isDense(@Name("node") Node node) {
...
+        if (limit <= 0)
+            throw new RuntimeException("limit must be a positive integer");
+
+        Map<String,Object> results = new HashMap<String, Object>();
+        HashMap<String, Object> parameters = new HashMap<>();
+
+        for( String label: labels) {
+            if(label.trim().length()>0) {
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: isEmpty()

File: src/main/java/apoc/nodes/Nodes.java (hunk 11 above)
Comment (jexp) [reply src/main/java/apoc/nodes/Nodes.java]: I'm not 100% sure apoc.periodic.iterate works in the same tx. please also test it in a server environment with a lot of highly connected labels (1M nodes 2M rels)

File: src/main/java/apoc/nodes/Nodes.java [hunk 16]
```diff
@@ -303,6 +304,26 @@ public boolean isDense(@Name("node") Node node) {
...
+                String query = String.format("call apoc.periodic.iterate( 'MATCH (n:`%s`) RETURN n',  'DETACH DELETE n', {batchsize:%d,iterateList:true, parallel:true}  )", label, limit);
+                db.execute(query, parameters);
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: you can either use the call without parameters or use them :)

File: src/main/java/apoc/nodes/Nodes.java [hunk 17]
```diff
@@ -303,6 +304,26 @@ public boolean isDense(@Name("node") Node node) {
...
+    @Procedure(mode = Mode.WRITE)
+    @Description("apoc.nodes.removewithlabels(['Person','Movie'],1000) | Will remove all the nodes having one of the given labels, via DETACH DELETE using apoc.periodic.iterate. Same limit used for each label. Default is 100. ")
+    public Stream<MapResult> removewithlabels(@Name("labels") List<String> labels, @Name(value = "limit",defaultValue="100")  Long limit) {
+        if (limit <= 0)
+            throw new RuntimeException("limit must be a positive integer");
+
+        Map<String,Object> results = new HashMap<String, Object>();
+        HashMap<String, Object> parameters = new HashMap<>();
+
+        for( String label: labels) {
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: stream to result one row per label

File: src/main/java/apoc/nodes/Nodes.java [hunk 18]
```diff
@@ -303,6 +304,26 @@ public boolean isDense(@Name("node") Node node) {
...
+            if(label.trim().length()>0) {
+                String query = String.format("call apoc.periodic.iterate( 'MATCH (n:`%s`) RETURN n',  'DETACH DELETE n', {batchsize:%d,iterateList:true, parallel:true}  )", label, limit);
+                db.execute(query, parameters);
+            }
+        }
+        MapResult mapResult = new MapResult(results);
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: you are not using the results, you should use the results of the iterate call as result for your method

inline

File: src/main/java/apoc/nodes/Nodes.java (hunk 17 above)
Comment (jexp) [reply src/main/java/apoc/nodes/Nodes.java]: `labels.stream().filter(empty).map(query)flatMap(execute).map(result)`

File: src/test/java/apoc/nodes/NodesTest.java [hunk 19]
```diff
@@ -27,7 +26,8 @@
     @Before
     public void setUp() throws Exception {
         db = TestUtil.apocGraphDatabaseBuilder().newGraphDatabase();
-        TestUtil.registerProcedure(db,Nodes.class);
+        TestUtil.registerProcedure(db,Nodes.class, Create.class);
+        TestUtil.registerProcedure(db, Periodic.class);
```
Comment (jexp) [inline src/test/java/apoc/nodes/NodesTest.java]: this is varargs, single line

File: src/test/java/apoc/nodes/NodesTest.java [hunk 20]
```diff
@@ -175,4 +175,184 @@ public void testDegreeDirectionOnly() {
         });
 
     }
+
+    @Test
+    public void testRemoveNodesEmptyLabels() {
+        String[] labels = {};
+        Map params = new HashMap();
```
Comment (jexp) [inline src/test/java/apoc/nodes/NodesTest.java]: Utils.map()

File: src/test/java/apoc/nodes/NodesTest.java [hunk 21]
```diff
@@ -175,4 +175,184 @@ public void testDegreeDirectionOnly() {
...
+        params.put("labels", labels);
+        params.put("limit", 50);
+        TestUtil.testResult(db,
+                "CALL apoc.nodes.removewithlabels",
```
Comment (jexp) [inline src/test/java/apoc/nodes/NodesTest.java]: don't use this call syntax, it's deprecated

File: src/test/java/apoc/nodes/NodesTest.java [hunk 22]
```diff
@@ -175,4 +175,184 @@ public void testDegreeDirectionOnly() {
...
+    public void testRemoveNodesWithLabels_AnEmptyLabel() {
+        String[] labels = {"aa", "bb" ,"","cc"};
+        Map params = new HashMap();
+        params.put("labels", labels);
+        params.put("limit", 50);
+        TestUtil.testResult(db,
+                "CALL apoc.nodes.removewithlabels",
+                params,
+                r -> {
+                    Map m = r.next();
```
Comment (jexp) [inline src/test/java/apoc/nodes/NodesTest.java]: generics

File: src/test/java/apoc/nodes/NodesTest.java [hunk 23]
```diff
@@ -175,4 +175,184 @@ public void testDegreeDirectionOnly() {
...
+                    Map map = (Map) m.get("value");
+                    assertNotNull(map);
```
Comment (jexp) [inline src/test/java/apoc/nodes/NodesTest.java]: really?!

Comment (jexp) [reply src/test/java/apoc/nodes/NodesTest.java]: please do some proper checking of the results, now that you have some

File: src/main/java/apoc/nodes/Nodes.java (hunk 11 above)
Comment (wadael) [reply src/main/java/apoc/nodes/Nodes.java]: I heard slotted is Entreprise only in your video. Unsure how it would behave in community.

Comment (wadael) [reply src/main/java/apoc/nodes/Nodes.java]: org.neo4j.graphdb.QueryExecutionException: Failed to invoke procedure  if I use runtime=slotted

As seen in debug:
community single [/home/jerome/OpenSource/wadael/neo4j-apoc-procedures/target/test-data/impermanent-db]

**Comment by wadael** [comment]
I think I cannot solve the not-a conflict

**Comment by mneedham** [comment]
@wadael closing this for now. Let us know if you'd like us to revisit

**Pull request closed without merging**